pkg/noderesourcetopology/least_numa.go

```diff
@@ -0,0 +1,192 @@
+					nRes[resName] = available
+				}
+			}
+		}
+	}
+}
+
+func numaNodesRequired(identifier string, qos v1.PodQOSClass, numaNodes NUMANodeList, resources v1.ResourceList) (bitmask.BitMask, int) {
+	combinationBitmask := bitmask.NewEmptyBitMask()
+	// we will generate combination of numa nodes from len = 1 to the number of numa nodes present on the machine
+	for i := 1; i <= len(numaNodes); i++ {
+		// generate combinations of len i
+		numaNodesCombination := combin.Combinations(len(numaNodes), i)
+		// iterate over combinations for given i
+		for _, combination := range numaNodesCombination {
+			// accumulate resources for given combination
+			combinationResources := combineResources(numaNodes, combination)
+
+			resourcesFit := true
+			onlyNonNUMAResources := true
+			for resource, quantity := range resources {
+				if quantity.IsZero() {
+					// why bother? everything's fine from the perspective of this resource
+					klog.V(4).InfoS("ignoring zero-qty resource request", "identifier", identifier, "resource", resource)
+					continue
+				}
+
+				combinationQuantity, ok := combinationResources[resource]
+				if !ok {
+					// non NUMA resource continue
+					continue
+				}
+
+				// there can be a situation where container/pod requests only non NUMA resources
+				onlyNonNUMAResources = false
+
+				if !isCombinationSuitable(qos, resource, quantity, combinationQuantity) {
+					resourcesFit = false
+					break
+				}
+
+			}
+			// if resources can be fit on given combination, just return the number of numa nodes requires to fit them
+			// according to TopologyManager if both masks are the same size pick the one that has less bits set
+			// https://github.com/kubernetes/kubernetes/blob/3e26e104bdf9d0dc3c4046d6350b93557c67f3f4/pkg/kubelet/cm/topologymanager/bitmask/bitmask.go#L146
+			// combin.Combinations is generating combinations in an order from the smallest to highest value
+			if resourcesFit {
+
+				// if a container/pod requests only non NUMA resources return empty bitmask and score of 0
+				if onlyNonNUMAResources {
+					return combinationBitmask, 0
+				}
+
+				combinationBitmask.Add(combination...)
+				return combinationBitmask, i
+			}
+		}
+	}
+
+	// score plugin should be running after resource filter plugin so we should always find sufficient amount of NUMA nodes
+	klog.V(1).InfoS("Shouldn't be here", "identifier", identifier)
+	return combinationBitmask, -1
+}
+
+func combineResources(numaNodes NUMANodeList, combination []int) v1.ResourceList {
+	resources := v1.ResourceList{}
+	for _, nodeIndex := range combination {
+		for resource, quantity := range numaNodes[nodeIndex].Resources {
+			if value, ok := resources[resource]; ok {
+				value.Add(quantity)
+				resources[resource] = value
+				continue
+			}
+			resources[resource] = quantity
+		}
+	}
+
+	return resources
+}
+
+func isCombinationSuitable(qos v1.PodQOSClass, resource v1.ResourceName, quantity, numaQuantity resource.Quantity) bool {
```

Comment:
this seems a copy-paste of `isNUMANodeSuitable`. I'm not against copy-paste per se ("a little duplication is better than a little dependency") but maybe we can evaluate other options? like rename/refactor the former and reuse it here.  Perhaps rename `isNUMANodeSuitable` to `isResourceSetSuitable` can fit and be clear enough in both call sites.

Reply:
Good idea.

Reply:
Done.